diff -up mesa-20100325/src/mesa/drivers/dri/intel/intel_fbo.c.da mesa-20100325/src/mesa/drivers/dri/intel/intel_fbo.c
--- mesa-20100325/src/mesa/drivers/dri/intel/intel_fbo.c.da	2010-03-25 08:42:47.000000000 +1000
+++ mesa-20100325/src/mesa/drivers/dri/intel/intel_fbo.c	2010-03-25 08:42:52.000000000 +1000
@@ -554,11 +554,6 @@ intel_render_texture(GLcontext * ctx,
        return;
    }
 
-   DBG("Begin render texture tid %lx tex=%u w=%d h=%d refcount=%d\n",
-       _glthread_GetID(),
-       att->Texture->Name, newImage->Width, newImage->Height,
-       irb->Base.RefCount);
-
    /* point the renderbufer's region to the texture image region */
    if (irb->region != intel_image->mt->region) {
       if (irb->region)
