diff -up mesa-20100325/configure.ac.no-mach64 mesa-20100325/configure.ac
--- mesa-20100325/configure.ac.no-mach64	2010-03-25 05:08:17.000000000 +1000
+++ mesa-20100325/configure.ac	2010-03-25 08:46:04.000000000 +1000
@@ -748,7 +748,7 @@ if test "$mesa_driver" = dri; then
             # the new interface.  i810 are missing because there is no
             # x86-64 system where they could *ever* be used.
             if test "x$DRI_DIRS" = "xyes"; then
-                DRI_DIRS="i915 i965 mach64 mga r128 r200 r300 r600 radeon \
+                DRI_DIRS="i915 i965 mga r128 r200 r300 r600 radeon \
                     savage tdfx unichrome swrast"
             fi
             ;;
@@ -756,13 +756,13 @@ if test "$mesa_driver" = dri; then
             # Build only the drivers for cards that exist on PowerPC.
             # At some point MGA will be added, but not yet.
             if test "x$DRI_DIRS" = "xyes"; then
-                DRI_DIRS="mach64 r128 r200 r300 r600 radeon tdfx swrast"
+                DRI_DIRS="r128 r200 r300 r600 radeon tdfx swrast"
             fi
             ;;
         sparc*)
             # Build only the drivers for cards that exist on sparc`
             if test "x$DRI_DIRS" = "xyes"; then
-                DRI_DIRS="mach64 r128 r200 r300 r600 radeon swrast"
+                DRI_DIRS="r128 r200 r300 r600 radeon swrast"
             fi
             ;;
         esac
@@ -779,7 +779,7 @@ if test "$mesa_driver" = dri; then
         fi
 
         if test "x$DRI_DIRS" = "xyes"; then
-            DRI_DIRS="i810 i915 i965 mach64 mga r128 r200 r300 r600 radeon tdfx \
+            DRI_DIRS="i810 i915 i965 mga r128 r200 r300 r600 radeon tdfx \
                 unichrome savage sis swrast"
         fi
         ;;
@@ -798,7 +798,7 @@ if test "$mesa_driver" = dri; then
 
     # default drivers
     if test "x$DRI_DIRS" = "xyes"; then
-        DRI_DIRS="i810 i915 i965 mach64 mga r128 r200 r300 r600 radeon \
+        DRI_DIRS="i810 i915 i965 mga r128 r200 r300 r600 radeon \
             savage sis tdfx unichrome swrast"
     fi
 
