diff -up mesa-20090612/src/mesa/drivers/osmesa/Makefile.dma mesa-20090612/src/mesa/drivers/osmesa/Makefile
--- mesa-20090612/src/mesa/drivers/osmesa/Makefile.dma	2009-06-12 13:57:29.000000000 +1000
+++ mesa-20090612/src/mesa/drivers/osmesa/Makefile	2009-06-12 13:57:42.000000000 +1000
@@ -39,7 +39,7 @@ default: $(TOP)/$(LIB_DIR)/$(OSMESA_LIB_
 # -DCHAN_BITS=16/32.
 $(TOP)/$(LIB_DIR)/$(OSMESA_LIB_NAME): $(OBJECTS) $(CORE_MESA)
 	$(MKLIB) -o $(OSMESA_LIB) -linker '$(CC)' -ldflags '$(LDFLAGS)' \
-		-major $(MESA_MAJOR) -minor $(MESA_MINOR) -patch $(MESA_TINY) \
+		-major 6 -minor 5 -patch 4 \
 		-install $(TOP)/$(LIB_DIR) $(MKLIB_OPTIONS) \
 		-id $(INSTALL_LIB_DIR)/lib$(OSMESA_LIB).$(MESA_MAJOR).dylib \
 		$(OSMESA_LIB_DEPS) $(OBJECTS) $(CORE_MESA)
