diff -up Mesa-9.0.3/src/gallium/drivers/nv50/codegen/nv50_ir.cpp.jx Mesa-9.0.3/src/gallium/drivers/nv50/codegen/nv50_ir.cpp
--- Mesa-9.0.3/src/gallium/drivers/nv50/codegen/nv50_ir.cpp.jx	2013-02-17 14:28:36.000000000 -0500
+++ Mesa-9.0.3/src/gallium/drivers/nv50/codegen/nv50_ir.cpp	2013-03-05 09:27:35.140943844 -0500
@@ -715,7 +715,7 @@ Instruction::clone(ClonePolicy<Function>
 {
    if (!i)
       i = new_Instruction(pol.context(), op, dType);
-#ifndef NDEBUG // non-conformant assert, so this is required
+#if 0
    assert(typeid(*i) == typeid(*this));
 #endif
 
